File: skills/comprehensiveSummarySkill/src/comprehensivesummary_skill/validation.py
from __future__ import annotations

from collections import Counter
from pathlib import Path
import re
from typing import Any

from bs4 import BeautifulSoup

from .core import _fact_tokens, normalize_language, word_count


def _finding(severity: str, code: str, message: str, details: Any = None) -> dict[str, Any]:
    item: dict[str, Any] = {"severity": severity, "code": code, "message": message}
    if details is not None:
        item["details"] = details
    return item


def validate_summary(
    source_path: Path,
    target_path: Path,
    *,
    intended_source: Path | None = None,
    expected_language: str | None = None,
) -> dict[str, Any]:
    source_path = source_path.expanduser().resolve()
    target_path = target_path.expanduser().resolve()
    source = BeautifulSoup(source_path.read_text(encoding="utf-8"), "html.parser")
    target = BeautifulSoup(target_path.read_text(encoding="utf-8"), "html.parser")
    findings: list[dict[str, Any]] = []

    marker = target.find("meta", attrs={"name": "summary-generator"})
    if marker is None or marker.get("content") != "comprehensivesummary-skill":
        findings.append(_finding("error", "ownership-marker", "Summary ownership metadata is missing."))
    source_lang = expected_language or (str(source.html.get("lang", "")) if source.html else "")
    target_lang = str(target.html.get("lang", "")) if target.html else ""
    try:
        if normalize_language(source_lang) != normalize_language(target_lang):
            findings.append(_finding("error", "language", "Summary language metadata differs from the source."))
    except Exception:
        findings.append(_finding("error", "language", "Source or summary has invalid language metadata."))

    article = target.find("article", attrs={"data-summary-body": True})
    if article is None:
        findings.append(_finding("error", "summary-body", "The main summary article is missing."))
        actual_words = 0
        main_text = ""
    else:
        clone = BeautifulSoup(str(article), "html.parser")
        main_text = clone.get_text(" ", strip=True)
        actual_words = word_count(main_text)
        if not clone.find("h1") or not clone.find("h2") or not clone.find_all("p"):
            findings.append(_finding("error", "semantic-shape", "Summary requires a title, thematic headings, and paragraphs."))
        for paragraph in clone.find_all("p"):
            if not paragraph.get_text(" ", strip=True):
                findings.append(_finding("error", "empty-paragraph", "Summary contains an empty paragraph."))
            if "dek" not in paragraph.get("class", []) and (
                not str(paragraph.get("data-clusters", "")).strip()
                or not str(paragraph.get("data-source-units", "")).strip()
            ):
                findings.append(_finding("error", "paragraph-traceability", "A summary paragraph lacks source tracing."))

    def meta_int(name: str) -> int | None:
        tag = target.find("meta", attrs={"name": name})
        try:
            return int(str(tag.get("content"))) if tag else None
        except (TypeError, ValueError):
            return None

    minimum, maximum, declared, words_per_minute = (
        meta_int("summary-minimum-words"), meta_int("summary-maximum-words"),
        meta_int("summary-actual-words"), meta_int("summary-words-per-minute"),
    )
    if minimum is None or maximum is None or declared is None or words_per_minute is None:
        findings.append(_finding("error", "word-metadata", "Word-budget metadata is incomplete."))
    else:
        if declared != actual_words:
            findings.append(_finding("error", "word-count", "Declared and actual summary word counts differ.", {"declared": declared, "actual": actual_words}))
        if not minimum <= actual_words <= maximum:
            findings.append(_finding("error", "word-budget", "Summary is outside the requested reading-time range.", {"minimum": minimum, "maximum": maximum, "actual": actual_words}))

    source_text = source.get_text(" ", strip=True)
    unsupported = sorted(_fact_tokens(main_text) - _fact_tokens(source_text))
    if unsupported:
        findings.append(_finding("error", "unsupported-factual-token", "Summary contains numbers, dates, URLs, emails, or DOI values absent from the source.", unsupported))

    source_map = target.find("details", attrs={"data-source-map": True})
    if source_map is None or source_map.find("table") is None:
        findings.append(_finding("error", "source-map", "The collapsible source map is missing."))
    else:
        for link in source_map.find_all("a", href=True):
            href = str(link["href"])
            if re.match(r"^[A-Za-z][A-Za-z0-9+.-]*:", href) or href.startswith(("#", "/", "//")):
                continue
            path_part = href.split("#", 1)[0].split("?", 1)[0]
            base = intended_source.parent if intended_source else target_path.parent
            local = (target_path.parent / path_part).resolve()
            if intended_source and Path(path_part).name == intended_source.name:
                local = intended_source
            if not local.is_file():
                findings.append(_finding("error", "source-link", f"Source-map link does not resolve: {href}"))

    ids = [str(tag["id"]) for tag in target.find_all(attrs={"id": True})]
    duplicates = [value for value, count in Counter(ids).items() if count > 1]
    if duplicates:
        findings.append(_finding("error", "duplicate-ids", "Summary contains duplicate element IDs.", duplicates))
    if target.find("script"):
        findings.append(_finding("error", "script", "Standalone summaries must not contain scripts."))
    paragraphs = [] if article is None else [
        " ".join(tag.get_text(" ", strip=True).casefold().split())
        for tag in article.find_all("p") if "dek" not in tag.get("class", [])
        if word_count(tag.get_text(" ", strip=True)) >= 8
    ]
    paragraph_counts = Counter(paragraphs)
    repeated = sum(count - 1 for count in paragraph_counts.values() if count > 1)
    if repeated:
        findings.append(_finding("error", "duplicate-paragraphs", "Summary contains duplicate substantial paragraphs.", {"count": repeated}))

    status = (
        "failed" if any(item["severity"] == "error" for item in findings)
        else "passed_with_warnings" if findings else "passed"
    )
    return {
        "status": status,
        "findings": findings,
        "metrics": {
            "actualWords": actual_words,
            "minimumWords": minimum,
            "maximumWords": maximum,
            "wordsPerMinute": words_per_minute,
            "estimatedMinutes": round(actual_words / words_per_minute, 2) if words_per_minute else None,
            "sourceMapRows": len(source_map.find_all("tr")) - 1 if source_map else 0,
            "unsupportedFactualTokens": len(unsupported),
            "duplicateParagraphs": repeated,
        },
        "limitations": (
            "Source tracing and deterministic checks cannot guarantee perfect factual or "
            "interpretive accuracy; semantic fidelity still depends partly on LLM review."
        ),
    }

File: skills/comprehensiveSummarySkill/src/validation.mjs
import fs from 'node:fs/promises';
import path from 'node:path';
import {
  parse,
  elements,
  first,
  named,
  has,
  meta,
  textOf,
  compact
} from './dom.mjs';
import {
  normalizeLanguage,
  wordCount,
  factTokens,
  exists,
  resolve
} from './core.mjs';
export async function validateSummary(
  sourcePath,
  targetPath,
  { intendedSource, expectedLanguage } = {}
) {
  sourcePath = resolve(sourcePath);
  targetPath = resolve(targetPath);
  const source = parse(await fs.readFile(sourcePath, 'utf8')),
    target = parse(await fs.readFile(targetPath, 'utf8')),
    findings = [];
  const finding = (code, message, details) =>
    findings.push({
      severity: 'error',
      code,
      message,
      ...(details === undefined ? {} : { details })
    });
  if (meta(target, 'summary-generator') !== 'comprehensivesummary-skill')
    finding('ownership-marker', 'Summary ownership metadata is missing.');
  try {
    if (
      normalizeLanguage(
        expectedLanguage || named(source, 'html')?.attribs.lang || ''
      ) !== normalizeLanguage(named(target, 'html')?.attribs.lang || '')
    )
      finding('language', 'Summary language metadata differs from the source.');
  } catch {
    finding('language', 'Source or summary has invalid language metadata.');
  }
  const article = first(
    target,
    n => n.name === 'article' && has(n, 'data-summary-body')
  );
  let mainText = '',
    actualWords = 0;
  if (!article) finding('summary-body', 'The main summary article is missing.');
  else {
    mainText = textOf(article);
    actualWords = wordCount(mainText);
    if (!named(article, 'h1') || !named(article, 'h2') || !named(article, 'p'))
      finding(
        'semantic-shape',
        'Summary requires a title, thematic headings, and paragraphs.'
      );
    for (const p of elements(article, n => n.name === 'p')) {
      if (!textOf(p))
        finding('empty-paragraph', 'Summary contains an empty paragraph.');
      if (
        !(p.attribs.class ?? '').split(/\s+/).includes('dek') &&
        (!(p.attribs['data-clusters'] ?? '').trim() ||
          !(p.attribs['data-source-units'] ?? '').trim())
      )
        finding(
          'paragraph-traceability',
          'A summary paragraph lacks source tracing.'
        );
    }
  }
  const integer = name => {
    const v = meta(target, name);
    return typeof v === 'string' && /^[+-]?\d+$/.test(v.trim())
      ? Number(v)
      : null;
  };
  const minimum = integer('summary-minimum-words'),
    maximum = integer('summary-maximum-words'),
    declared = integer('summary-actual-words'),
    wpm = integer('summary-words-per-minute');
  if ([minimum, maximum, declared, wpm].includes(null))
    finding('word-metadata', 'Word-budget metadata is incomplete.');
  else {
    if (declared !== actualWords)
      finding('word-count', 'Declared and actual summary word counts differ.', {
        declared,
        actual: actualWords
      });
    if (actualWords < minimum || actualWords > maximum)
      finding(
        'word-budget',
        'Summary is outside the requested reading-time range.',
        { minimum, maximum, actual: actualWords }
      );
  }
  const sourceFacts = factTokens(textOf(source)),
    unsupported = [...factTokens(mainText)]
      .filter(t => !sourceFacts.has(t))
      .sort();
  if (unsupported.length)
    finding(
      'unsupported-factual-token',
      'Summary contains numbers, dates, URLs, emails, or DOI values absent from the source.',
      unsupported
    );
  const map = first(
    target,
    n => n.name === 'details' && has(n, 'data-source-map')
  );
  if (!map || !named(map, 'table'))
    finding('source-map', 'The collapsible source map is missing.');
  else
    for (const a of elements(map, n => n.name === 'a' && has(n, 'href'))) {
      const href = a.attribs.href;
      if (/^[A-Za-z][A-Za-z0-9+.-]*:|^[#/]/.test(href)) continue;
      const part = href.split('#')[0].split('?')[0];
      let local = path.resolve(path.dirname(targetPath), part);
      if (
        intendedSource &&
        path.basename(part) === path.basename(intendedSource)
      )
        local = intendedSource;
      if (!(await exists(local)))
        finding('source-link', `Source-map link does not resolve: ${href}`);
    }
  const ids = elements(target, n => has(n, 'id')).map(n => n.attribs.id),
    duplicates = [...new Set(ids.filter((id, i) => ids.indexOf(id) !== i))];
  if (duplicates.length)
    finding(
      'duplicate-ids',
      'Summary contains duplicate element IDs.',
      duplicates
    );
  if (named(target, 'script'))
    finding('script', 'Standalone summaries must not contain scripts.');
  const paragraphs = article
    ? elements(
        article,
        n =>
          n.name === 'p' &&
          !(n.attribs.class ?? '').split(/\s+/).includes('dek') &&
          wordCount(textOf(n)) >= 8
      ).map(n =>
        compact(
          textOf(n)
            .toLowerCase()
            .replaceAll('ß', 'ss')
        )
      )
    : [];
  const repeated = paragraphs.length - new Set(paragraphs).size;
  if (repeated)
    finding(
      'duplicate-paragraphs',
      'Summary contains duplicate substantial paragraphs.',
      { count: repeated }
    );
  return {
    status: findings.length ? 'failed' : 'passed',
    findings,
    metrics: {
      actualWords,
      minimumWords: minimum,
      maximumWords: maximum,
      wordsPerMinute: wpm,
      estimatedMinutes: wpm
        ? Math.round((actualWords / wpm) * 100) / 100
        : null,
      sourceMapRows: map ? elements(map, n => n.name === 'tr').length - 1 : 0,
      unsupportedFactualTokens: unsupported.length,
      duplicateParagraphs: repeated
    },
    limitations:
      'Source tracing and deterministic checks cannot guarantee perfect factual or interpretive accuracy; semantic fidelity still depends partly on LLM review.'
  };
}
